### DISCUSSION TITLE
Drum module connectivity

### DISCUSSION THREAD

**Discussion by leoinlios** [jamulussoftware/jamulus#2102] [General]
Hi guys, I carry out 3 tasks with my drums, 1) Jamulus, 2) DAW recording, 3) Play drums. I currently don't have an interface and will purchase one JUST so that I can follow Jamulus' Golden rule. My connectivity will therefore be as follows:
![image](https://user-images.githubusercontent.com/4964741/139667971-7f105ccb-afe0-4b0a-8198-e6bbba361d18.png)

It is quite annoying that I need to have 2 headphones (1 for the interface and 1 for the drum module) depending on the task at hand or that I need to unplug/replug the 1 headphone which is obviously not recommended as I will end up weakening the connections. 

Am I missing something here, do you guys do it differently? @gilgongo ?

I thought that I could get 2 extra analogue cables and connect the interface back to the drum module and always have the headsets on the drum module but this means that when I play drums I will have to have the interface ON, which is redundant. I have added numbers and letters to facilitate the dialogue. 

Thank you for your help
L

**Comment by gilgongo** [comment]
Yes it's not ideal I guess, but in my case my audio interface (a UMC202HD) is powered over USB so I can't not have it on when then computer is running, and never use the audio interface in my drum module because I can't mute my output from it. So I just hit the "direct monitor" button on the UMC202 to listen to my computer's sound when I'm not in Jamulus. 

Mind you, I prefer to use a pair of wireless headphones for "casual" use with my computer (not Jamulus for obvious latency reasons), so do in fact have two sets for that reason.

**Comment by pljones** [comment]
You could just have the one set up and use it for everything, of course.

Drum Module out -> PC In (via onboard or external interface)
PC In -> DAW Input
DAW Input -> DAW Send to Jamulus
Jamulus Input -> the world -> Jamulus Output
Jamulus Output -> DAW Receive from Jamulus
DAW Receive from Jamulus -> DAW Master Out
DAW Master Out -> PC Out
PC Out (via onboard or external interface) -> headphones

When you're using Jamulus, make sure that "DAW Input" _only_ routes to "DAW Send to Jamulus".

When you're recording, route "DAW Input" wherever you want.

When you're just playing, route "DAW Input" to "DAW Master Out"

Yes, it does mean turning the computer on always...

It's fairly easy, also, to automate switching between those three routing set ups (at least in Reaper and presumably other DAWs with MIDI or OSC automation), if you've got an Android or iOS phone or tablet to hand.

**Reply by leoinlios** [reply]
@pljones , I would be really keen on getting your solution to work as this means not getting an additional interface, just for using Jamulus. **However:** When I set the DAW (Cakewalk) driver mode to ASIO,..
![image](https://user-images.githubusercontent.com/4964741/139716321-ffb639e0-4ad2-437d-8bf2-7270ce626356.png)
I only see inputs and outputs of the Drum module (TD-27), the PC or Jamulus are not an option.  
![image](https://user-images.githubusercontent.com/4964741/139716822-635b2d53-4a7c-4a54-8f5a-dd7f0489074d.png)
Switching the driver mode to anything but ASIO will of course affect latency really badly. 
I therefore cannot see anything about "DAW send to Jamulus" anywhere.
In addition if I have the DAW open, which is connected via ASIO to my drum module, Jamulus does not start as it complaints:
![image](https://user-images.githubusercontent.com/4964741/139717334-de78713d-ce91-4c60-8b56-e1cf31718503.png)
So how can I tell Jamulus to take input from the DAW or output to the DAW? 
When I open Jamulus when the DAW is closed, there is no other input/output (on ASIO) but the drum module:
![image](https://user-images.githubusercontent.com/4964741/139717853-6d9b683c-33cd-4097-addf-35104070ed49.png)

so for me it has only ever been using the DAW or using Jamulus. 

What's the trick for using both?

**Reply by pljones** [reply]
Most ASIO drivers only allow one application to use them at a time.

So what you need an an application to "own" the ASIO driver that then allows multiple other applications to connect to it.

There are a few ways:
- JACK for Windows with JackRouter (which many people have got to work, just not me - probably due to my Windows install being an upgrade from Windows 7 through 8 to 10 on a 2013 built PC with various old drivers installed causing it pain; JackRouter, itself, needs a text file editing as Windows Administrator to configure it)
  - JackRouter is needed for other ASIO applications - they see JackRouter providing ASIO connections like any other ASIO driver.
  - For Jamulus, you'd then be able to use the Jamulus for Jack for Windows version (and we'd be interested to hear your experiences with it).
- Reaper with ReaRoute (which isn't free, although it's cheap - it's what I use)
  - Other applications then connect to the ReaRoute ASIO connections.
- One of the other "virtual ASIO cabling" solutions -- most of these seem less reliable / introduce more latency than JACK or Reaper, though, from what I've heard

So you set up that appliction to handle your "real" ASIO driver, then any other applications use the offered "virtual" ASIO connections to connect to that application.

**Reply by leoinlios** [reply]
Makes sense - thanks for taking the time to explain. I tried Voicemeeter Banana to do the very thing you described. Problem is my Jamulus latency went **from** ping 10ms delay 30ms **to** ping 10ms delay 50ms (I set samples to the lowest, which are much higher than what ASIO direct allows me to set), but 50ms is unplayable for me. For background, my PC: DESKTOP gaming Win 10 PC (16GB ram - SSD - i7 6700 processor @3.4GHz - 4GB GPU), latest Windows Updates. 

I hope that a real external interface will not have this dramatic negative impact on my latency, in fact I hope it reduces the latency compared to connecting my instrument directly to the PC via ASIO....I find it hard to believe though. I will find out when I get the interface tomorrow.

**Reply by pljones** [reply]
Voicemeeter Banana falls into that third category -- I've really not heard of anything worth using there.

You might want to give JACK a try.  For me, even after a clean up, it still fails (I'm fairly sure down to the version of one of the libraries it uses that doesn't like my sound card - which is ancient, and no longer supported).

**Reply by gilgongo** [reply]
@leoinlios The "JACK aware" verison of Jamulus, that is: https://github.com/jamulussoftware/jamulus/releases/download/r3_8_1/jamulus_3.8.1_win_jack.exe

**Reply by leoinlios** [reply]
Thanks both

**Comment by leoinlios** [comment]
Hi @pljones  / @gilgongo  - I purchased an interface and as the sound is far worse than what I used to hear when connecting my drum module directly to Jamulus (Roland TD-27 via ASIO), I started to explore the JACK option recommended above. My objectives: 
**1)** For Jamulus to receive my computer usb mic and my drum module sounds, 
**2)** for the Jamulus output to come out to the PC headphone (and not back to the drum module as it does not let me switch off direct monitoring). 
Here is the connection so far:
![image](https://user-images.githubusercontent.com/4964741/140050954-b69ec233-adad-4987-b11b-bad79057e3a0.png)

With this setup I achieved objective 1, and I get great latency. I followed the steps here: https://jamulus.io/kb/2020/11/24/Multiple-Audio-Interfaces.html. I know that objective 1 can be done much simpler just using Voicemeeter but the latency is terrible (per above comment). 

Is there a way to achieve objective 2, such as routing Jamulus back into Voicemeeter? I tried in JACK connecting the OUTs of Jamulus to the INs of Voicemeeter but Voicemeeter cannot see Jamulus as an available input for Input 2. 

Is this possible at all?

Thanks

**Reply by npostavs** [reply]
> Is there a way to achieve objective 2, such as routing Jamulus back into Voicemeeter? I tried in JACK connecting the OUTs of Jamulus to the INs of Voicemeeter but Voicemeeter cannot see Jamulus as an available input for Input 2.

Did you try connecting the OUTs of Jamulus to the system playback_1,2?

**Reply by leoinlios** [reply]
I believe System is my drum module. I need the output from Jamulus to be the PC, not the drum module, per above.

**Reply by jujudusud** [reply]
In the image, you use JackRouter as sound device and this is not good. You have to install and you the JACK for windows Jamulus version that Gilgongo point in this comment : https://github.com/jamulussoftware/jamulus/discussions/2102#discussioncomment-1572594

**Reply by npostavs** [reply]
> I believe System is my drum module.

Hmm, you mean JACK isn't seeing your other devices at all? I thought it should be able to, but I'm not very familiar with it, so perhaps that's just expected.

> Voicemeeter cannot see Jamulus as an available input for Input 2.

Maybe you need to select JackRouter as the Input 2; since Jamulus goes through JACK, Voicemeeter might not see Jamulus as a separate device. Also, I think you might also need to use the banana version of Voicemeeter, because the non-banana version you are using doesn't allow sending output to either A1 or A2 individually (I believe you get both versions installed by default, there's a different shortcut for the banana version).

**Comment by jujudusud** [comment]
Why do you use Voicemeeter ?
Using JACK allow you to stop using voicemeeter

**Reply by leoinlios** [reply]
Please can you explain how you can use JACK alone to send to Jamulus both your PC microphone AND your drum module, and also how you use JACK to output Jamulus to the PC headphone?

I went with the posted instructions: https://jamulus.io/kb/2020/11/24/Multiple-Audio-Interfaces.html

Thank you

**Reply by pljones** [reply]
Do not use Windows system devices at all.  If you use them, you need ASIO4ALL as a minimum.  Voicemeeter does not remove latency in the Windows audio subsystems.  Neither does ASIO4ALL - but it seems to introduce the least.  They'll still not perform as well as a hardware interface with an external mic and headphones.  The hardware interface should also allow your Roland audio out to be plugged in, too.  Then you'd only have one ASIO device to worry about.

**Reply by jujudusud** [reply]
> Please can you explain how you can use JACK alone to send to Jamulus both your PC microphone AND your drum module, and also how you use JACK to output Jamulus to the PC headphone?

It is not possible to use multiple audio devices simultanneously on windows with JACK. As It is also not possible with ASIO exclusive mode. That is a limitation needed by the fact that we want to achieve the lowest latency.

**Reply by leoinlios** [reply]
ok thanks for clarifying. I thought this was possible as this article https://jamulus.io/kb/2020/11/24/Multiple-Audio-Interfaces.html says "otherwise you can use voicemeeter to redirect the sound to another device, which I will not cover on this guide"

**Reply by pljones** [reply]
> "otherwise you can use voicemeeter to redirect the sound to another device, which I will not cover on this guide"

There's a reason for that: it's a bad idea and introduces unacceptable latency... :)
